lib/src/arithmetic/values/floating_point_values/floating_point_value.dart

```diff
@@ -310,19 +338,101 @@ class FloatingPointValue implements Comparable<FloatingPointValue> {
+                mantissa.toBigInt().toDouble() /
+                    pow(2.0, mantissa.width - (explicitJBit ? 1 : 0))) *
             pow(2.0, exponent.toInt() - bias);
       }
     }
     return doubleVal;
   }
 
+  /// Convert to a [FloatingPointValue] with a mantissa that is one smaller
+  /// due to the implicit J-bit.
+  FloatingPointValue toFloatingPointValue() {
+    final norm = canonicalize();
+    return FloatingPointValue(
+        sign: norm.sign,
+        exponent: norm.exponent,
+        mantissa: norm.mantissa.getRange(0, norm.mantissa.width - 1));
+  }
+
   /// Return a Logic true if this FloatingPointVa;ie contains a normal number,
   /// defined as having mantissa in the range `[1,2)`.
-  bool isNormal() => exponent != LogicValue.ofInt(0, exponent.width);
+  bool isNormal() {
+    if (explicitJBit) {
+      final e = exponent.toInt();
+      final m = mantissa.toInt();
+      final int normMantissa;
+      if (e < mantissa.width) {
+        normMantissa = 1 << (mantissa.width - e - 1);
+      } else {
+        normMantissa = 1;
+      }
+      return (e > 0) && (m >= normMantissa);
+    } else {
+      return exponent != LogicValue.ofInt(0, exponent.width);
+    }
+  }
+
+  /// Check if the mantissa and exponent stored are compatible
+  bool isLegalValue() {
+    if (explicitJBit) {
+      final e = exponent.toInt();
+      final m = mantissa.toInt();
+      // TODO(desmonddak): We need to check this with bit-pattern testing
+      // of legal mantissas and that exponents are compatible with those.
+      // Basically, if e > 0 then we expect a 1 somewhere.  If e == 0 then
+      // we expect anything except a leading 1 in the mantissas.
+      final normMantissa = 1 << (mantissa.width - 1);
+
+      return ((e == 0) && (m < normMantissa)) || ((e > 0) && (m >= 1));
+    }
+    return true;
+  }
+
+  /// Return the cananocalized form of [FloatingPointValue] which
+  /// has the leading 1 at the front of the mantissa, or further right if
+  /// subnormal.
+  FloatingPointValue canonicalize() {
+    if (explicitJBit) {
+      var expVal = exponent.toInt();
+      var mant = mantissa;
+      if (!isAnInfinity) {
+        if (!isNaN) {
+          if (mant.or() == LogicValue.one) {
+            while ((mant[-1] == LogicValue.zero) & (expVal > 1)) {
+              expVal--;
+              mant = mant << 1;
+            }
+            if ((mant[-1] == LogicValue.zero) & (expVal == 1)) {
+              // Make canonical: if it cannot be made normal, it is subnormal
+              expVal = 0;
+            } else if ((mant[-1] == LogicValue.one) & (expVal == 0)) {
+              expVal = 1;
+            }
+          } else {
+            expVal = 0;
+          }
+        } else {
+          return populator(
+                  exponentWidth: exponentWidth,
+                  mantissaWidth: mantissaWidth,
+                  explicitJBit: explicitJBit)
+              .nan;
+        }
+      }
+      return FloatingPointValue(
```

Comment:
should this be using populator instead?

Reply:
This is a pattern we use in each converter, right ?  (ofString, ofDouble, ofBigInts, random, etc).

Reply:
Ahhh I missed it in this spot.